Add first draft of 4.4-dev4 article (#952)
godotengine/godot-website: collections/_article/dev-snapshot-godot-4-4-dev-4.md

Git, to preserve your projects in case of corruption or data loss.

[Jump to the **Downloads** section](#downloads), and give it a spin right now, or continue reading to learn more about improvements in this release. You can also try the [**Web editor**](https://editor.godotengine.org/releases/4.4.dev4/), [**XR editor**](https://www.meta.com/experiences/godot-game-engine/7713660705416473/), or the **Android editor** for this release (join the [Android editor testing group](https://groups.google.com/g/godot-testers) to get access to pre-release builds).

-----

*The cover illustration is from* [****](), *a ! It is developed in Godot 4.3 by [](). You can wishlist the game [on Steam]() and follow the developers on [Twitter]().*

## Highlights

In case you missed them, see the [4.4 dev 1](/article/dev-snapshot-godot-4-4-dev-1/), [dev2](/article/dev-snapshot-godot-4-4-dev-2/),
and [dev 3](/article/dev-snapshot-godot-4-4-dev-3/) release notes for an overview of some key features which were
already in that snapshot, and are therefore still available for testing in dev 3.

Here are a few highlights that you might find particularly exciting!

> ### Embed Game Window

Reviewer: I'd make the title more clear:

```suggestion
### First steps towards an embedded game window with interactive in-game editing
```

Author: I agree with changing this. I went with "Interactive in-game editing" to keep it simple, the explanation below explains that it is a first step to the fully embedded game window

One of the most requested features is the ability to debug the running game from within the editor including pausing the
game, clicking on elements within the scene, and embedding the running game window inside the editor. Dev4 comes with

> the first two of these workflows thanks to the contribution from [YeldhamDev]( https://github.com/YeldhamDev) of Lone Wolf

Reviewer: ```suggestion
the first two of these workflows thanks to the contribution from [YeldhamDev](https://github.com/YeldhamDev) of Lone Wolf
```

happy with.

“Ubershaders” are a version of each shader that contains all features and are compiled at load time. We pair the
ubershader with the ability to compile specialized shader pipelines in the background. This way when we draw a new
object, instead of freezing and waiting for the pipeline to compile, we fallback on the ubershader and compile the
specialized shader in the background.

Most games will notice the improvement from ubershaders without any content changes. However, there are certain cases
where the ubershader won't work and some content changes will be needed to avoid shader stutter. However, the
workarounds for avoiding shader stutter are now very simple and are well documented already in the [official docs](https://docs.godotengine.org/en/latest/tutorials/performance/pipeline_compilations.html).

Taking advantage of the new ubershader infrastructure, we can now optimize our shaders by creating specialized variants
that skip unused pathways. We have already begin doing so to optimize our existing shaders
([GH-98825](https://github.com/godotengine/godot/pull/98825)).

> ### Add shadow_caster_mask to Light3D

Reviewer: ```suggestion
### Add Shadow Caster Mask property to Light3D
```

This long awaited improvement from [EMBYRDEV](https://github.com/EMBYRDEV) allows users to apply a mask on Light3Ds to
select what rendering layers will be considered when casting shadows
([GH-85338](https://github.com/godotengine/godot/pull/85338)). Previously it was only possible to disable shadows from a
GeometryInstance (for all Light3Ds) or a Light3D (for all GeometryInstances). This allows much more fine-grained control
that allows users to further optimize dynamic lights and control where shadows appear in their games. 

### Always add decimal when converting float to string 

Currently when printing float values Godot trims trailing `.0`s as the are technically unnecessary. However, this could
lead to issues for users who need to distinguish between floating point numbers and integers.

> [KoBeWi](https://github.com/kobewi) has now changed the behavior when floats are converted to strings so that the
> trailing `.0` is always printed when floats are converted to strings ([GH-47502](https://github.com/godotengine/godot/pull/47502)). 

Reviewer 2: ```suggestion
[KoBeWi](https://github.com/kobewi) has now changed the behavior when floats are converted to strings to always add the decimal number, even if it would be `.0`.
This makes floats always recognisable when they are converted to strings ([GH-47502](https://github.com/godotengine/godot/pull/47502)). 
```
Or similar

Author: ```suggestion
[KoBeWi](https://github.com/kobewi) has now changed the behavior when floats are converted to strings to always add the decimal number, even if it would be `.0`.
This makes floats always recognizable when they are converted to strings ([GH-47502](https://github.com/godotengine/godot/pull/47502)). 
```

Looks good, small change to use US spelling instead of UK

This change in behavior may break your project if you are relying on the current serialization behavior. If this change
breaks your project, please let us know by opening an issue on Github. 

```gdscript

> # before
> var a: float = 1
> print(a)
> # prints 1

Reviewer 2: ```suggestion
# Before:
var a: float = 1.0
print(a)
# Prints 1
```

```

```gdscript

> # after
> var a: float = 1
> print(a)
> # prints 1.0

Reviewer 2: ```suggestion
# After:
var a: float = 1.0
print(a)
# Prints 1.0
```

- Buildsystem: CI: Add MinGW/GCC build to Windows GHA ([GH-97446](https://github.com/godotengine/godot/pull/97446)).
- Core: Implement array based hash map AHashMap ([GH-92554](https://github.com/godotengine/godot/pull/92554)).
- Core: Expose `Geometry2D.bresenham_line()` method ([GH-74714](https://github.com/godotengine/godot/pull/74714)).
- Core: Add Object support for String.format ([GH-65962](https://github.com/godotengine/godot/pull/65962)).
- Dotnet: Change generated `On{SignalName}` to `EmitSignal{SignalName}` ([GH-97588](https://github.com/godotengine/godot/pull/97588)).
- Dotnet: Generate `ReadOnlySpan<T>` Overloads for GodotSharp APIs ([GH-96329](https://github.com/godotengine/godot/pull/96329)).
- Editor: Improve UI of sun angle in the sun/preview env popup ([GH-97747](https://github.com/godotengine/godot/pull/97747)).
- Editor: Add submenu support to EditorContextMenuPlugin ([GH-97292](https://github.com/godotengine/godot/pull/97292)).
- Editor: Improve display for lightmap probes in the editor ([GH-83863](https://github.com/godotengine/godot/pull/83863)).

> - Editor: Add missing audio bus button styles #([GH-3443 ](https://github.com/godotengine/godot/pull/3443 )).

Reviewer 3: ```suggestion
- Editor: Add missing audio bus button styles #([GH-93443](https://github.com/godotengine/godot/pull/93443)).
```
